accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -621,6 +614,488 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+  var x;
+  while(b) { // x has no initialization inferred type
+  }
+}
+```
+
+
+At a non-back edge join point in the program, it is an error if a non-`late`
+**untyped variable** has an **initialization inferred type** on one of the
+reachable paths to the join point and not on another reachable path.  Note that
+it is valid for a variable to have no **initialization inferred type** on a path
+into a join point when the join point is known to be unreachable along that
+path.  It is recommended that tools signal this error by indicating that the
+variable has not been assigned a value on all paths.
+
+```dart
+void test(bool b) {
+  var x; // x is an untyped variable
+  if (b) {
+    x = 3;
+  } // Error, x is not assigned on all paths
+
+  late final y; // y is a late untyped variable
+  if (b) {
+    y = 3;
+  } // Ok, y has inferred type `int`
+
+  final z;
+  if (b) {
+     z = 3;
+  } else {
+    return;
+  }  // Ok, z has inferred type `int` even though it is unassigned on one path.
+}
+```
+
+At non-back edge join points, some adjustment to the **initialization inferred
+type** is done to account for nullability.  Informally, a non-`late` **untyped**
+variable must be assigned a value on every path, but a variable declared `late`
+may be assigned a value on only some of the paths into a join.  At a join point,
+the **initialization inferred type** on each path that has one must agree,
+except that they may differ on nullability, in which case they are adjusted as
+described below.  In all cases, the values assigned must all have the same type
+(up to nullability), or type `Null` - otherwise the analysis fails to infer a
+type for the variable and it is an error.  More precisely, the **initialization
+inferred type** of a variable after a join point is specified as follows.
+
+At a non-backed edge join point in the program, if a non-`late` **untyped
+variable** has one or more *reachable* paths into the join on which it has no
+**initialization inferred type**, and one or more *reachable* paths into the
+join on which it has an **initialization inferred type**, it is an error and
+inference fails for the variable. In other words, a non-`late` variable must
+either be assigned a value on every reachable path or on no reachable paths.
+
+At a non-back edge join point in the program, if an **untyped variable** has no
+**initialization inferred type** on every path into the join, then it has no
+**initialization inferred type** after the join.
+
+If neither of the above cases apply, then there is at least one path (reachable
+or not) into the join which has an **initialization inferred type**, and the
+**initialization inferred type**  for the variable after the join point
+is computed as follows:
+  - If any path into the join (reachable or not) has a **final initialization
+    inferred type** `T`, then:
+    - It is an error if any other path (reachable or not) has a **final
+      initialization type** `S` and `S` is not equal to `T`.
```

Comment:
Is "equal" defined in terms of normalization?  E.g. is `FutureOr<int?>` "equal" to `FutureOr<int?>?` for purposes of this check?